File: frontend/src/features/active-directory/utils/ldapAttributeValueTypes.js
const LDAP_SUPPORTED_VALUE_TYPES =
  Object.freeze([
    'single_text',
    'boolean',
    'integer32',
    'integer64',
  ])

const LDAP_INTEGER_BOUNDS =
  Object.freeze({
    integer32: Object.freeze({
      minimum: -2147483648,
      maximum: 2147483647,
    }),
    integer64: Object.freeze({
      minimum: Number.MIN_SAFE_INTEGER,
      maximum: Number.MAX_SAFE_INTEGER,
    }),
  })

function normalizeLdapValueType(valueType) {
  const normalized = String(
    valueType || ''
  )
    .trim()
    .toLowerCase()

  if (
    !LDAP_SUPPORTED_VALUE_TYPES.includes(
      normalized
    )
  ) {
    throw new Error(
      `Type de valeur LDAP non pris en charge : ${
        normalized || 'inconnu'
      }.`
    )
  }

  return normalized
}

function parseLdapTypedInputValue(
  valueType,
  rawValue
) {
  const normalizedType =
    normalizeLdapValueType(valueType)

  if (normalizedType === 'single_text') {
    if (typeof rawValue !== 'string') {
      throw new Error(
        'Une chaîne de caractères est obligatoire.'
      )
    }

    return rawValue
  }

  if (normalizedType === 'boolean') {
    if (typeof rawValue === 'boolean') {
      return rawValue
    }

    const normalized = String(
      rawValue ?? ''
    )
      .trim()
      .toLowerCase()

    if (normalized === 'true') {
      return true
    }

    if (normalized === 'false') {
      return false
    }

    throw new Error(
      'Une valeur booléenne est obligatoire.'
    )
  }

  if (
    typeof rawValue === 'number' &&
    Number.isSafeInteger(rawValue)
  ) {
    return rawValue
  }

  if (typeof rawValue !== 'string') {
    throw new Error(
      'Une valeur entière est obligatoire.'
    )
  }

  const normalized = rawValue.trim()

  if (!/^-?\d+$/.test(normalized)) {
    throw new Error(
      'Une valeur entière est obligatoire.'
    )
  }

  const numericValue = Number(normalized)

  if (!Number.isSafeInteger(numericValue)) {
    throw new Error(
      'Cet entier dépasse la précision sûre de JavaScript.'
    )
  }

  return numericValue
}

function normalizeLdapTypedEditorValue({
  valueType,
  value,
  minLength = 0,
  maxLength = null,
  minValue = null,
  maxValue = null,
}) {
  const normalizedType =
    normalizeLdapValueType(valueType)

  if (normalizedType === 'single_text') {
    if (typeof value !== 'string') {
      throw new Error(
        'Une chaîne de caractères est obligatoire.'
      )
    }

    const normalizedValue = value.trim()

    if (!normalizedValue) {
      throw new Error(
        'La valeur ne peut pas être vide avec l’opération set.'
      )
    }

    if (
      normalizedValue.includes(
        String.fromCharCode(0)
      ) ||
      normalizedValue.includes('\r') ||
      normalizedValue.includes('\n')
    ) {
      throw new Error(
        'La valeur contient un caractère interdit.'
      )
    }

    if (
      minLength > 0 &&
      normalizedValue.length < minLength
    ) {
      throw new Error(
        'La valeur est trop courte.'
      )
    }

    if (
      Number.isInteger(maxLength) &&
      normalizedValue.length > maxLength
    ) {
      throw new Error(
        'La valeur est trop longue.'
      )
    }

    return normalizedValue
  }

  if (normalizedType === 'boolean') {
    if (typeof value !== 'boolean') {
      throw new Error(
        'Une valeur booléenne est obligatoire.'
      )
    }

    return value
  }

  if (!Number.isSafeInteger(value)) {
    throw new Error(
      'Une valeur entière sûre est obligatoire.'
    )
  }

  const hardBounds =
    LDAP_INTEGER_BOUNDS[normalizedType]

  const effectiveMinimum =
    Number.isSafeInteger(minValue)
      ? Math.max(
          hardBounds.minimum,
          minValue
        )
      : hardBounds.minimum

  const effectiveMaximum =
    Number.isSafeInteger(maxValue)
      ? Math.min(
          hardBounds.maximum,
          maxValue
        )
      : hardBounds.maximum

  if (
    effectiveMinimum >
    effectiveMaximum
  ) {
    throw new Error(
      'Les bornes numériques sont incohérentes.'
    )
  }

  if (value < effectiveMinimum) {
    throw new Error(
      'La valeur est inférieure au minimum autorisé.'
    )
  }

  if (value > effectiveMaximum) {
    throw new Error(
      'La valeur dépasse le maximum autorisé.'
    )
  }

  return value
}

function formatLdapTypedEditorValue(
  valueType,
  value
) {
  if (
    value === null ||
    value === undefined ||
    value === ''
  ) {
    return '—'
  }

  const normalizedType =
    normalizeLdapValueType(valueType)

  if (normalizedType === 'boolean') {
    return value ? 'Oui' : 'Non'
  }

  return String(value)
}

export {
  LDAP_INTEGER_BOUNDS,
  LDAP_SUPPORTED_VALUE_TYPES,
  formatLdapTypedEditorValue,
  normalizeLdapTypedEditorValue,
  normalizeLdapValueType,
  parseLdapTypedInputValue,
}

File: api/app/services/ldap_attribute_value_types.py
from __future__ import annotations

from dataclasses import asdict, dataclass


LDAP_SUPPORTED_VALUE_TYPES = frozenset({
    "single_text",
    "boolean",
    "integer32",
    "integer64",
})

_INTEGER_BOUNDS = {
    "integer32": (
        -2147483648,
        2147483647,
    ),
    "integer64": (
        -9223372036854775808,
        9223372036854775807,
    ),
}


@dataclass(frozen=True)
class LDAPValueNormalizationResult:
    valid: bool
    normalized_value: str | int | bool | None
    errors: tuple[dict[str, str], ...]

    def to_dict(self) -> dict:
        payload = asdict(self)
        payload["errors"] = [
            dict(item)
            for item in self.errors
        ]
        return payload


def _error(
    code: str,
    message: str,
) -> dict[str, str]:
    return {
        "code": code,
        "message": message,
    }


def normalize_ldap_typed_value(
    *,
    value_type: str,
    value: object,
    minimum_length: int = 0,
    maximum_length: int | None = None,
    minimum_value: int | None = None,
    maximum_value: int | None = None,
) -> LDAPValueNormalizationResult:
    errors: list[dict[str, str]] = []
    normalized_value: str | int | bool | None = None

    if value_type not in LDAP_SUPPORTED_VALUE_TYPES:
        errors.append(
            _error(
                "unsupported_value_type",
                "Type de valeur LDAP non pris en charge.",
            )
        )

    elif value_type == "single_text":
        if not isinstance(value, str):
            errors.append(
                _error(
                    "invalid_value_type",
                    "Valeur texte obligatoire.",
                )
            )
        else:
            normalized_value = value.strip()

            if not normalized_value:
                errors.append(
                    _error(
                        "empty_set_value",
                        "La valeur texte ne peut pas être vide.",
                    )
                )

            if any(
                char in normalized_value
                for char in (
                    "\x00",
                    "\r",
                    "\n",
                )
            ):
                errors.append(
                    _error(
                        "forbidden_control_character",
                        "Caractère de contrôle interdit.",
                    )
                )

            if len(normalized_value) < minimum_length:
                errors.append(
                    _error(
                        "value_too_short",
                        "Valeur trop courte.",
                    )
                )

            if (
                maximum_length is not None
                and len(normalized_value) > maximum_length
            ):
                errors.append(
                    _error(
                        "value_too_long",
                        "Valeur trop longue.",
                    )
                )

    elif value_type == "boolean":
        if type(value) is not bool:
            errors.append(
                _error(
                    "invalid_value_type",
                    "Valeur booléenne obligatoire.",
                )
            )
        else:
            normalized_value = value

    else:
        if type(value) is not int:
            errors.append(
                _error(
                    "invalid_value_type",
                    "Valeur entière obligatoire.",
                )
            )
        else:
            normalized_value = value
            hard_minimum, hard_maximum = (
                _INTEGER_BOUNDS[value_type]
            )

            effective_minimum = (
                hard_minimum
                if minimum_value is None
                else max(
                    hard_minimum,
                    minimum_value,
                )
            )

            effective_maximum = (
                hard_maximum
                if maximum_value is None
                else min(
                    hard_maximum,
                    maximum_value,
                )
            )

            if effective_minimum > effective_maximum:
                errors.append(
                    _error(
                        "invalid_numeric_bounds",
                        "Bornes numériques incohérentes.",
                    )
                )
            else:
                if value < effective_minimum:
                    errors.append(
                        _error(
                            "value_below_minimum",
                            "Valeur inférieure au minimum.",
                        )
                    )

                if value > effective_maximum:
                    errors.append(
                        _error(
                            "value_above_maximum",
                            "Valeur supérieure au maximum.",
                        )
                    )

    return LDAPValueNormalizationResult(
        valid=not errors,
        normalized_value=normalized_value,
        errors=tuple(errors),
    )
